test/lib/PlacesTestUtils.js

```diff
@@ -24,22 +26,69 @@ const PlacesTestUtils = Object.freeze({
    *            [optional] visitDate: visit date in microseconds from the epoch
    *            [optional] referrer: nsIURI of the referrer for this visit
    *          }
-   *
-   * @return {Promise}
-   *            resolves When all visits have been added successfully.
-   *            rejects JavaScript exception.
    */
   addVisits: Task.async(function*(placeInfo) {
-    let promise = new Promise((resolve, reject) => {
-      let places = [];
-      if (placeInfo instanceof Ci.nsIURI) {
-        places.push({uri: placeInfo});
-      } else if (Array.isArray(placeInfo)) {
-        places = places.concat(placeInfo);
-      } else {
-        places.push(placeInfo);
+    let places = [];
+
+    if (placeInfo instanceof Ci.nsIURI) {
+      places.push({uri: placeInfo});
+    } else if (Array.isArray(placeInfo)) {
+      places = places.concat(placeInfo);
+    } else {
+      places.push(placeInfo);
+    }
+
+    // Keep track of the number of rows for final validation
+    function* getPlacesRows() {
+      let conn = yield PlacesUtils.promiseDBConnection();
+      let rows = yield conn.execute("SELECT id, url, title, frecency FROM moz_places");
+      return rows;
+    }
+    let initialNumRows = yield getPlacesRows().length;
+
+    // setup listeners for history events to validate db inserts and to ensure
+    // that eventual queries will lead to expected results
+    let urlList = places.map(place => place.uri.spec);
+    let urlSet = new Set(urlList);
+    let urlCount = {};
+    for (let url of urlList) {
+      urlCount[url] = 0;
+    }
+    // there are 3 notifications per place inserted
+    let totalCount = urlList.length * 3;
+    let notifCount = 0;
+
+    let notifPromise = new Promise(resolve => {
+      function collectNotifs(aURI) {
+        if (urlSet.has(aURI.spec)) {
```

Comment:
Q: if `urlSet.has(aURI.spec)` doesn't have the URL, don't we already have bigger issues (i.e., something is screwing the testing environment with either a frecency change or title change)? Maybe we should report that problem and stop?